ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/AggregatingAttestationPoolV2.java

```diff
@@ -0,0 +1,446 @@
+          attestationData.getBeaconBlockRoot(),
+          attestationData.getTarget().getRoot());
+      return Optional.empty();
+    }
+
+    final Bytes dataHash = attestationData.hashTreeRoot();
+
+    // Atomically get or create the set for the slot
+    dataHashBySlot
+        // Use ConcurrentHashMap::newKeySet for a concurrent set implementation
+        .computeIfAbsent(attestationData.getSlot(), __ -> ConcurrentHashMap.newKeySet())
+        .add(dataHash);
+
+    // Atomically get or create the MatchingDataAttestationGroupV2
+    final MatchingDataAttestationGroup attestationGroup =
+        attestationGroupByDataHash.computeIfAbsent(
+            dataHash,
+            __ ->
+                new MatchingDataAttestationGroupV2(
+                    spec, attestationData, committeesSize)); // Pass spec, data, committeesSize
+
+    return Optional.of(attestationGroup);
+  }
+
+  // This method primarily reads immutable/thread-safe data or state likely protected
+  // by RecentChainData's own synchronization. It should be safe without external locks here.
+  private Optional<Int2IntMap> getCommitteesSizeUsingTheState(
+      final AttestationData attestationData) {
+    // we can use the first state of the epoch to get committees for an attestation
+    final MiscHelpers miscHelpers = spec.atSlot(attestationData.getSlot()).miscHelpers();
+    final Optional<UInt64> maybeEpoch = recentChainData.getCurrentEpoch();
+    // the only reason this can happen is we don't have a store yet.
+    if (maybeEpoch.isEmpty()) {
+      return Optional.empty();
+    }
+    final UInt64 currentEpoch = maybeEpoch.get();
+    final UInt64 attestationEpoch = miscHelpers.computeEpochAtSlot(attestationData.getSlot());
+
+    LOG.debug("currentEpoch {}, attestationEpoch {}", currentEpoch, attestationEpoch);
+    if (attestationEpoch.equals(currentEpoch)
+        || attestationEpoch.equals(currentEpoch.minusMinZero(1))) {
+
+      return recentChainData
+          .getBestState()
+          .flatMap(
+              state -> {
+                try {
+                  // Assuming getBeaconCommitteesSize is thread-safe or state access is handled
+                  return Optional.of(
+                      spec.getBeaconCommitteesSize(
+                          state.getImmediately(), attestationData.getSlot()));
+                } catch (IllegalStateException e) {
+                  LOG.debug(
+                      "Couldn't retrieve state for committee calculation of slot {}",
+                      attestationData.getSlot());
+                  return Optional.empty();
+                }
+              });
+    }
+
+    // attestation is not from the current or previous epoch
+    // this is really an edge case because the current or previous epoch is at least 31 slots
+    // and the attestation is only valid for 64 slots, so it may be epoch-2 but not beyond.
+    final UInt64 attestationEpochStartSlot = miscHelpers.computeStartSlotAtEpoch(attestationEpoch);
```

Comment:
i've been thinking about this a little.... i think if we only ever use the start slot of the current epoch or start slot of previous epoch, then that's probably safest (because of limited cache largely)
we've got 2 options, i think simplest is 
 - like above, if attestation is current or previous epoch use bestState
 - otherwise use first slot of previous epoch (which generally is cached still)

if we can easily log cache misses here it may be a good thing...

i'd probably avoid doing the start slot of the attestation epoch because that may be 2 epochs behind (per comment) and almost certainly require a regeneration from ... probably the justified state...